<a href="https://colab.research.google.com/github/jes147/Kelompok-A-CaseStudy-Data-Wrangling/blob/main1/Kelompok_A_CaseStudy_Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><h1><strong><font color="blue">Studi Kasus Data Wrangling</font></strong></h1></center>

![taudata Analytics](https://github.com/taudataid/eLearning/raw/master/images/covers/cover_ui-n-taudata.jpg "taudata Analytics")

Anggota Kelompok:
1. Annisa Zahra (2006463295)
2. Safira Mecca Lovani (2006486550)
3. Annisa Fitria Anwar Damanik (2106639900)
4. Medifa Puspaningrum (2106634944)
5. Jeremi Okto Ricardo Marpaung (2006568733)
6. Angelica Patricia Djaya Saputra (2006522000)
7. Jesica Michele Yulianto (2106725002)

In [25]:
import warnings; warnings.simplefilter('ignore')
import pandas as pd

try: # Running Locally, yakinkan "file_" berada di folder "data"
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')
except: # Running in Google Colab
    !mkdir data
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_dosen.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_kelas.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_krs.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mahasiswa.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mata_kuliah.csv
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')

mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape

((14, 4), (4, 6), (4, 4), (7, 6), (15, 8), (9, 3), (33, 7))

In [26]:
# Code untuk Soal No 01 diberikan sebagai Contoh dan Bonus
# Pelajari dengan baik

nilai_dosen = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen = nilai_dosen.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_dosen = pd.merge(dosen, nilai_dosen, left_on='NIDN', right_index=True)
nilai_dosen = nilai_dosen[['NIDN', 'nama', 'mean_nilai']]
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_dosen = nilai_dosen.set_index(['NIDN']) 
nilai_dosen = pd.concat([nilai_dosen, sb], axis=1)
nilai_dosen

nama  mean_nilai  std_nilai
NIDN                                      
123    Andri Basuki   85.750000   7.741355
234        zulfikar   80.428571  15.883504
345   Annisa Fitria   60.090909  16.220077
787            yudi   70.571429  13.010984

# Interpretasi 01

Berdasarkan tabel di atas, Dosen Andri Basuki memiliki rata-rata nilai 85.75 namun memiliki standar deviasi yang paling rendah diantara dosen lainnya yaitu 7.74. Sehingga nilai yang diberikan Dosen Andri Basuki kurang bervariasi dan cenderung mendekati mean daripada dosen lainnya.

Sedangkan, untuk Dosen Annisa Fitria memiliki mean paling rendah yaitu 60 dan standar deviasi paling tinggi yaitu 16.22. Didapatkan hasil interpretasi bahwa Dosen Annisa Fitria cenderung memberikan nilai yang lebih bervariasi dibandingkan dosen lainnya namun dengan catatan nilai yang diberikan berada di sekitar 60 (karena std masih tergolong kecil)


In [27]:
# Code Untuk Soal No 02
nd_df = pd.merge(dosen, nilai, on='NIDN')
nd_df = nd_df[['Pendidikan_Terakhir','nilai']]
mean_nilai = nd_df.groupby(nd_df['Pendidikan_Terakhir']).mean()[['nilai']]
mean_nilai = mean_nilai.rename({'nilai':'mean_nilai'}, axis=1)
std_nilai = nd_df.groupby(nd_df['Pendidikan_Terakhir']).std()[['nilai']]
std_nilai = std_nilai.rename({'nilai':'std_nilai'}, axis=1)
nd_df = pd.concat([mean_nilai, std_nilai], axis=1)
nd_df

mean_nilai  std_nilai
Pendidikan_Terakhir                       
S2                    64.166667  15.561075
S3                    83.266667  12.068061

# Interpretasi 02:

Dari data terlihat bahwa rata - rata pemberian nilai dosen dengan pendidikan terakhir S3 lebih tinggi yaitu sebesar 83.266667 daripada dosen dengan pendidikan terakhir S2 yaitu sebesar 64.166667. Selain itu, dapat dilihat pula dari standar deviasinya, nilai yang diberikan oleh dosen dengan pendidikan terakhir S2 lebih tersebar dibandingkan dengan nilai yang diberikan oleh dosen dengan pendidikan terakhir S3. Dapat disimpulkan bahwa terdapat perbedaan dalam pemberian nilai antara dosen dengan pendidikan terakhir S3 dan S2. 

In [28]:
# Code Soal No 03

nilai_mahasiswa = nilai.groupby(nilai['nim']).mean()[['nilai']]
nilai_mahasiswa = nilai_mahasiswa.rename({'nilai': 'mean_nilai'}, axis=1)
nilai_mahasiswa = pd.merge(mahasiswa, nilai_mahasiswa, left_on='nim', right_index=True)
nilai_mahasiswa = nilai_mahasiswa[['nim', 'Asal_Sekolah', 'mean_nilai']]
sb = nilai.groupby(nilai['nim']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1)
nilai_mahasiswa = nilai_mahasiswa.set_index(['nim'])
nilai_mahasiswa = pd.concat([nilai_mahasiswa, sb], axis=1)
nilai_mahasiswa.pivot_table(index= 'Asal_Sekolah')

mean_nilai  std_nilai
Asal_Sekolah                       
MA             71.937500  19.673045
SMK            85.900000   4.693409
SMU            66.416667  16.729164

# Interpretasi 03:

Dari data yang dihasilkan, terlihat bahwa rata-rata nilai mahasiswa yang berasal dari SMK lebih tinggi dari rata-rata nilai mahasiswa yang berasal dari SMU dan MA, yaitu 85.900000. Sedangkan rata-rata nilai mahasiswa yang paling rendah berasal dari SMU , yaitu 66.41667. Selain itu, dapat dilihat pula dari std nya, yang mana nilai rentangnya cukup jauh sehingga nilai mahasiswa rentangnya dapat dikatakan cukup bervariasi.


**Maka dapat disimpulkan bahwa terdapat perbedaan performa akademik antara mahasiswa yang berasal dari MA, SMK ataupun SMU. Yang mana dapat dilihat pula bahwa performa akademik terbaik berasal dari mahasiswa lulusan SMK, yang disusul dengan mahasiswa MA, lalu mahasiswa SMU.**

In [29]:
# Code Soal No 04

nilai_mahasiswa_berdasarkan_gender = nilai.groupby(mahasiswa['nim']).mean()[['nilai']]
# mencari mean dari tiap mahasiswa
nilai_mahasiswa_berdasarkan_gender = nilai_mahasiswa_berdasarkan_gender.rename({'nilai': 'mean_nilai'}, axis=1)
# merge data
nilai_mahasiswa_berdasarkan_gender = pd.merge(mahasiswa, nilai_mahasiswa_berdasarkan_gender, left_on='nim', right_index=True)
# tampilkan dataset dengan fitur nim, gender, dan mean 
nilai_mahasiswa_berdasarkan_gender = nilai_mahasiswa_berdasarkan_gender[['nim', 'Gender', 'mean_nilai']]
nilai_mahasiswa_berdasarkan_gender = nilai_mahasiswa_berdasarkan_gender.set_index(['nim'])
nilai_mahasiswa_berdasarkan_gender

Gender  mean_nilai
nim                       
194015    Pria        70.0
194001    Pria        80.0
194002    Pria        85.0
194003    Pria        90.0
194004    Pria        92.0
194005    Pria        91.0
194006  Wanita        93.0
194007  Wanita        85.0
194008  Wanita        60.0
194009    Pria        63.0
194010  Wanita        62.0
194011  Wanita        61.0
194012    Pria        70.0
194013    Pria        90.0
194014  Wanita        88.0

In [30]:
# rata rata nilai berdasarkan gender
nilai_mahasiswa_berdasarkan_gender.pivot_table(index='Gender')

mean_nilai
Gender            
Pria     81.222222
Wanita   74.833333

# Interpretasi 04

Di sini, kami menggunakan nilai rata rata pada tiap mahasiswa yang ada. Pada data terdapat perbedaan kemampuan akademik antara laki-laki terhadap perempuan. Rata-rata nilai pada gender laki-laki (pria) = 81.2222 dan perempuan (wanita) = 74.83333. Hal ini berarti kemampuan akademik laki-laki lebih baik dibandingkan perempuan.

In [31]:
#Code soal No 5

#merge dataframe nilai dan mahasiswa dari kolom nim
nilai_mahasiswa=pd.merge(nilai,mahasiswa, on='nim')
nilai_mahasiswa=nilai_mahasiswa[['Angkatan', 'id_mata_kuliah','nilai']]

#mencari mean
mean =nilai_mahasiswa.groupby(['Angkatan','id_mata_kuliah']).mean()[['nilai']]
mean =mean.rename({'nilai':'mean_nilai'}, axis=1)

#mencari standar deviasi
sb = nilai_mahasiswa.groupby(['Angkatan','id_mata_kuliah']).std()[['nilai']]
sb = sb.rename({'nilai':'std_nilai'}, axis=1)

#mencari min
min = nilai_mahasiswa.groupby(['Angkatan','id_mata_kuliah']).min()[['nilai']]
min = min.rename({'nilai':'min_nilai'}, axis=1)

#mencari max
max = nilai_mahasiswa.groupby(['Angkatan','id_mata_kuliah']).max()[['nilai']]
max = max.rename({'nilai':'max_nilai'}, axis=1)

nilai_mahasiswa = nilai_mahasiswa.set_index(['Angkatan'])
nilai_mahasiswa = pd.concat([mean, sb, min, max], axis=1)
nilai_mahasiswa

mean_nilai  std_nilai  min_nilai  max_nilai
Angkatan id_mata_kuliah                                             
2019     CS01             80.500000  17.253019         55         92
         KAL01            85.750000   7.741355         70         93
         KAL02            63.500000   8.103497         55         74
         STA02            54.000000  12.543258         40         70
2020     CS01             80.333333  17.616280         60         91
         KAL01            70.571429  13.010984         60         90
         STA02            63.666667  29.399546         31         88

# Interpretasi 05

Dari data terlihat bahwa secara keseluruhan, terdapat variasi perbedaan prestasi akademik yang bisa dilihat dari perbedaan nilai rata-rata mahasiswa angkatan 2019 dengan mahasiswa angkatan 2020.

Disini, kami membagi prestasi akademik tersebut per mata kuliah:
- Mata kuliah CS01, rata-rata nilai mahasiswa angkatan 2019 **lebih tinggi**, yakni 80,5 dengan nilai paling kecil 55 dan nilai paling besar 92; dibanding mahasiswa 2020, yakni 80,33 dengan nilai paling kecil 60 dan nilai paling besar 91.
- Mata kuliah KAL01, rata-rata nilai mahasiswa angkatan 2019 juga **lebih tinggi**, yakni 85,75 dengan nilai terkecil 70 dan nilai terbesar 93; dibanding mahasiswa 2020, yakni 70,57 dengan nilai terkecil 60 dan nilai terbesar 90.
- Mata kuliah STA02, rata-rata nilai mahasiswa angkatan 2019 **lebih rendah**, yakni 54 dengan nilai terkecil 40 dan nilai terbesar 70; dibanding mahasiswa 2020, yakni 63,67 dengan nilai terkecil 31 dan terbesar 88.
- Data mata kuliah KAL02 **tidak kami gunakan** dalam perbandingan karena tidak tersedianya data tersebut bagi mahasiswa angkatan 2020.
- Dari standar deviasi secara keseluruhan, diketahui bahwa data nilai mahasiswa angkatan 2020 lebih tersebar dibanding data nilai mahasiswa angkatan 2019.

In [32]:
# Code soal No 06

#1 (untuk 1 langsung dijelaskan di interpretasi 6)

#2
krs1 = krs.drop(columns='id')
krs1.head()

id_mata_kuliah kode_kelas     nim  tahun semester
0          KAL01      SIS1J  193001   2021    genap
1          STA01       D101  193001   2021    genap
2           OR01       D212  193001   2021    genap
3          MK002       P601  194001   2021    genap
4          MK002       P601  194001   2021    genap

In [33]:
#3
#Insight : Mencari Insight Nilai berdasarkan Asal Daerah
#1. Merge data mahasiswa dan nilai
mhsxnilai=pd.merge(mahasiswa,nilai,on='nim')
mhsxnilai.head()

#2. Cari nilai rata-rata per mahasiswa
meannilai = mhsxnilai.groupby(mhsxnilai['nama']).mean()[['nilai']]
meannilai = meannilai.rename({'nilai': 'mean_nilai'}, axis=1)
meannilai.head()

#3. Merge data mahasiswa dengan nilai rata-rata per mahasiswa
mhsxmeannilai=pd.merge(mahasiswa,meannilai,on='nama')
mhsxmeannilai.head()

#5. Mencari rata-rata nilai berdasarkan daerah
nilaixdaerah = mhsxmeannilai.groupby(mhsxmeannilai['Asal_Daerah']).mean()[['mean_nilai']]
nilaixdaerah.head()

#6. Cari std mahasiswa
stdnilai = mhsxnilai.groupby(mhsxnilai['nama']).mean()[['nilai']]
stdnilai = stdnilai.rename({'nilai': 'std_nilai'}, axis=1)
stdnilai.head()

#7. Merge data mahasiswa dengan std per mahasiswa
mhsxstdnilai=pd.merge(mahasiswa,stdnilai,on='nama')
mhsxstdnilai.head()

#8. Mencari std berdasarkan daerah
stdxdaerah = mhsxstdnilai.groupby(mhsxstdnilai['Asal_Daerah']).std()[['std_nilai']]
stdxdaerah.head()

tabel = pd.merge(nilaixdaerah,stdxdaerah, on='Asal_Daerah')
tabel

mean_nilai  std_nilai
Asal_Daerah                       
Lokal         76.233333  15.086915
Luar Daerah   70.725000   8.200838

# Interpretasi 06



1.   Dapat dilihat secara langung dari data, bahwa jumlah dosen hanya ada 4 sedangkan jumlah matkul ada 9. Sehingga kurang maksimal dalam mengolah data dikarenakan data yang diolah hanya sedikit dan tidak bisa mengambil kesimpulan secara general. Selain itu, jumlah mata kuliah untuk angkatan 2020 tidak sama dengan jumlah mata kuliah untuk angkatan 2019, sehingga mata kuliah tersebut tidak kami gunakan.
2.   Pada krs, jika dilakukan drop kolom 'id' dimana disuspect sebagai index, akan terdapat duplicated data (pada #2)
3.   Terdapat informasi baru atau insight, dimana nilai rata-rata dan standar deviasi dari mahasiswa dikelompokkkan berdasaran asal daerah. Dengan kesimpulan mahasiswa asal daerah lokal memiliki nilai rata-rata lebih tinggi dibanding mahasiswa asal luar daerah.



**REKOMENDASI**:
Kami merekomendasikan penambahan jumlah sample data agar kesimpulan yang diambil bisa lebih general.